README.md
# BagTest

Ventilators In Peoria (VIP)

Test the VIP fixture

This is written and tested on the Arduino Mega 2560

Serial Port Info:

stty -F /dev/ttyACM0 115200
cu -l /dev/ttyACM0 -s 115200

 ~/arduino-1.8.12/arduino --upload BagTest.ino


#To compile
make verify

#To compile and upload
make verify

#To compile and upload
make upload

#Start data collection
make test 

In another terminal 

tail -f out.csv




Plot/history.txt
 1024  git checkout 8a45a45507fa79e53ac6491853636ef4a25e2706
 1025  git rebase -i HEAD~2
 1026  git checkout master
 1027  git status
 1028  git rebase -i HEAD~2
 1029  git log --pretty=oneline --abbrev-commit
 1030  d9f1cf5 Changed again
 1031  105fd3d Changed content
 1032  df33c8a First commit
 1033  git log --pretty=oneline --abbrev-commit
 1034  git push origin +master
 1035  git status
 1036  gitk --all &
 1037  git status
 1038  gvim out.dat 
 1039  git status
 1040  git diff
 1041  git status
 1042  git add .
 1043  git commit
 1044  git reset --HARD
 1045  git reset HARD
 1046  git reset HEAD
 1047  git status
 1048  git reset
 1049  git status
 1050  git reset HEAD~`
 1051  git reset HEAD~1
 1052  git status
 1053  gvim out.dat 
 1054  git status
 1055  git add .
 1056  git commit
 1057  git push origin master
 1058  git status
 1059  git diff
 1060  git commit 
 1061  clear
 1062  ll
 1063  gvim out.dat 
 1064  git status
 1065  gvim outldat
 1066  gvim out.dat
 1067  git add
 1068  git add .
 1069  git commit
 1070  git push origin master 
 1071  arduino-1.8.12/arduino &
 1072  cd VENTILATOR/
 1073  ll
 1074  cd BagTest/
 1075  ll
 1076  gvim out.dat 
 1077  git status
 1078  git add .
 1079  git commit
 1080  git push origin master 
 1081  clear
 1082  git diff
 1083  git add .
 1084  git reset ~HEAD
 1085  git status
 1086  git reset 
 1087  git status
 1088  gvim out.dat 
 1089  git status
 1090  git add .
 1091  git status
 1092  git commit
 1093  git status
 1094  git push origin master 
 1095  clear
 1096  git status
 1097  pwd
 1098  ll
 1099  git diff
 1100  clear
 1101  gitk --all &
 1102  gvim out.dat 
 1103  git status
 1104  git add 
 1105  git add .
 1106  git commit
 1107  git push
 1108  git push origin master 
 1109  clear
 1110  ps auxw | grep cup
 1111  killall cups
 1112  killall cupsd
 1113  sudo service cupsd restart
 1114  sudo /etc/init.d/cups restart
 1115  gvim out.dat 
 1116  diff
 1117  git diff
 1118  git status 
 1119  git add .
 1120  git commit
 1121  git push origin master
 1122  clear
 1123  ssh dt
 1124  gvim .bashrc 
 1125  arduino-1.8.12/arduino &
 1126  gti status
 1127  cd VENTILATOR/BagTest/
 1128  ll
 1129  git diff
 1130  tail out.dat 
 1131  ping dt
 1132  clear
 1133  ll
 1134  bing dt
 1135  sudo bing dt
 1136  ping dt
 1137  clear
 1138  ll
 1139  ssh root@192.168.42.1
 1140  ping dt
 1141  clear
 1142  gvim out.dat 
 1143  git status
 1144  git add .
 1145  git commit
 1146  git status
 1147  git push origin master
 1148  gitk --all &
 1149  find ~ | grep -i Cialas 
 1150  find ~ | grep -i ods
 1151  gvim out.dat 
 1152  t 
 1153  git stat
 1154  git status
 1155  git add .
 1156  git commit
 1157  git status
 1158  gitk --all 
 1159  gvim out.dat 
 1160  git statuus
 1161  git status
 1162  git add .
 1163  git commit
 1164  git push origin master
 1165  gvim out.dat 
 1166  git status
 1167  git add out.dat 
 1168  git status
 1169  git commit
 1170  git push origin master 
 1171  clear
 1172  git status
 1173  git add out.dat 
 1174  ll
 1175  git status
 1176  gvim out.dat 
 1177  git status
 1178  git add out.dat 
 1179  ll
 1180  git status
 1181  git commit
 1182  git push origin master
 1183  gvim out.dat 
 1184  git statu
 1185  git sta
 1186  git stat
 1187  git status
 1188  git add .
 1189  git status
 1190  git commit 
 1191  git push origin master
 1192  git status
 1193  git add .
 1194  git commit
 1195  gvim out.dat 
 1196  git add .
 1197  git status
 1198  git commit
 1199  git push origin master 
 1200  git status
 1201  git add .
 1202  git commit
 1203  clear
 1204  git push origin master 
 1205  gvim out.dat 
 1206  git commit
 1207  git add 
 1208  git add .
 1209  git commit 
 1210  git push origin
 1211  clear
 1212  git status
 1213  clear
 1214  cd ..
 1215  cd ROBS_LIGHT_TOWER/
 1216  ll
 1217  mkdir REPO
 1218  cd REPO
 1219  git clone https://github.com/degnarraer/Arduino.git
 1220  git 
 1221  ll
 1222  cd Arduino/
 1223  ll
 1224  git fetch --tags 
 1225  git pull
 1226  cd LightTower/
 1227  ll
 1228  grep -i timer *
 1229  gvim ADCSampler.cpp 
 1230  cd ..
 1231  grep -r TC_CMR_WAVE 
 1232  cd ..
 1233  cd arduino-1.8.12/
 1234  grep -r TC_CMR_WAVE 
 1235  grep -r TC0 *
 1236  grep -r TC1 *
 1237  grep -r TC2 *
 1238  grep -r TC2 * | grep for
 1239  grep -r TC1 * | grep for
 1240  grep -r TC1 * | grep For
 1241  grep -r TC1 * | grep -v form | grep for
 1242  grep -r TC1 * 
 1243  grep TC_FOR_TIMER -r 
 1244  grep TC2 -r
 1245  grep TC3 -r
 1246  grep TC4 -r
 1247  clear
 1248  ll
 1249  cd ../ROBS_LIGHT_TOWER/
 1250  ll
 1251  cd REPO/
 1252  ll
 1253  cd Arduino/
 1254  ll
 1255  grep -r TC
 1256  grep -r TC0
 1257  grep -r TC1
 1258  grep -r TC2
 1259  grep -r TC3
 1260  find | grep Hop
 1261  find | grep EZ
 1262  ssh 192.168.1.197 -X -Y 
 1263  ssh pi@192.168.1.197 -X -Y 
 1264  ll
 1265  cd TIMELAPSE/
 1266  ll
 1267  mv timelapse.mp4 timelapse_ON_xps15_before.mp4
 1268  clear
 1269  ll
 1270  scp pi@192.168.1.197:~/TIMELAPSE/timelapse.mp4 .
 1271  nautilus .
 1272  scp pi@192.168.1.197:~/TIMELAPSE/timelapse.mp4 .
 1273  nautilus . &
 1274  ll
 1275  scp pi@192.168.1.197:~/TIMELAPSE/timelapse.mp4 .
 1276  nautilus .& 
 1277  ~/arduino-1.8.12/arduino &
 1278  scp pi@192.168.1.197:~/TIMELAPSE/timelapse.mp4 .
 1279  cd /etc/
 1280  killall nouveau
 1281  ps auxw  | nov
 1282  ps auxw  | grep  nov
 1283  ll
 1284  cd /etc/init.d/
 1285  ll
 1286  grep touch *
 1287  grep mouse *
 1288  clear
 1289  ssh pi@192.168.1.197 -X -Y 
 1290  ll
 1291  cd arduino-1.8.12/
 1292  ll
 1293  ./arduino
 1294  j
 1295  ll
 1296  ssh pi@192.168.1.133 -X -Y 
 1297  cd TIMELAPSE/
 1298  ll
 1299  dmesg
 1300  j
 1301  scp pi@192.168.1.197:~/TIMELAPSE/timelapse.mp4 .
 1302  nautilus . &
 1303  ssh -XY pi@192.168.1.133
 1304  ll
 1305  cd ../Downloads/
 1306  ll
 1307  ll -rt 
 1308  sudo apt install xz-utils 
 1309  man xz
 1310  sudo apt install xz-utils 
 1311  man xz
 1312  unxz arduino-1.8.12-linux

fs6122.cpp
#include "fs6122.h"
#include <Wire.h>
#include <Streaming.h>  // cout <iostream> functionality using Serial << endl;

void fs6122_init(){
  Wire.begin();

  //Configure filter depth
  Wire.beginTransmission(FS6122_ADDRESS);
  Wire.write(FS6122_CMD_WRITE_FILTER_DEPTH);
  Wire.write(byte(FS6122_FILTER_DEPTH));
  Wire.endTransmission();
}

void fs6122_zeroFlowCal(){
  // Must be done when there is guarnteed zero flow in the pipe
  // This function blocks for up to a second, and should not be done in periodic tasking
  Wire.beginTransmission(FS6122_ADDRESS);
  Wire.write(FS6122_CMD_CAL_FLOW_RATE);
  Wire.write(byte(0xFF)); //arbitrary byte
  Wire.endTransmission();
  delay(300);

  Wire.beginTransmission(FS6122_ADDRESS);
  Wire.write(FS6122_CMD_CAL_PRESSURE);
  Wire.write(byte(0xFF)); //arbitrary byte
  Wire.endTransmission();
  delay(300);
}

void fs6122_readSmlpM_umH2O(pressure_flow_type *pf){
   long reading = 0;

   Wire.beginTransmission(FS6122_ADDRESS);
   Wire.write(FS6122_CMD_READ_FLOW_RATE_AND_PRESSURE);
   Wire.endTransmission();

   Wire.requestFrom(FS6122_ADDRESS, 9); 

   if (8 <= Wire.available()) { // if two bytes were received
      for(int byteIdx = 0; byteIdx < 4; byteIdx++){
         pf->mSLPM = pf->mSLPM << 8;    // shift high byte to be high 8 bits
         pf->mSLPM |= Wire.read(); // receive low byte as lower 8 bits
      }

      if (4 <= Wire.available()) {
         for(int byteIdx = 0; byteIdx < 4; byteIdx++){
            pf->umH2O = pf->umH2O << 8;    // shift high byte to be high 8 bits
            pf->umH2O |= Wire.read(); // receive low byte as lower 8 bits
         }
      }

   } else {
      //ERROR - sensor did not return data. TODO - return something meaningful
       pf->mSLPM = -42;
       pf->umH2O = -42;
   }
}


float fs6122_readSmlpM(){
  long reading = 0;

  Wire.beginTransmission(FS6122_ADDRESS);
  Wire.write(FS6122_CMD_READ_FLOW_RATE);
  Wire.endTransmission();

  Wire.requestFrom(FS6122_ADDRESS, 4); 
  
  if (4 <= Wire.available()) { // if two bytes were received
    for(int byteIdx = 0; byteIdx < 4; byteIdx++){
      reading = reading << 8;    // shift high byte to be high 8 bits
      reading |= Wire.read(); // receive low byte as lower 8 bits
    }
    return (float)reading;
  } else {
    //ERROR - sensor did not return data. TODO - return something meaningful
    return -42;
  }
}


DATA/plotBagTest.py
"""
Created on Sat Jul 4 14:51:36 2020

@author: dgupta
"""
import argparse
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

class plotBagTest():
    """ Reading Data from Greg's BAG Test csv file and plot the columns vs time(s)"""
    def __init__(self, filen, title=None):
        self.filen = filen
        self.title = title
        self.head = ['millis', 'mSLPM', 'umH2O', 'cpuLoad', 'position',
                'temperature' , 'direction', 'pwmSpeed', 'analogRead']
        if title==None:
            self.title = self.filen
        
    def loadfile(self, skiprows=8):
        df = pd.read_csv(self.filen, skiprows=skiprows, header=None, 
            names=self.head)
        df.index = (df.iloc[:,0] - df.iloc[0,0])/1000
        df.index.name = 'sec'
        return df
    
    def run(self): #, png=False):
        df = self.loadfile()
        df.plot(figsize=(10,12), subplots=True, grid=True, title=self.title);
        # if png:
        #     plt.savefig('out.png', bbox_inches='tight')
        plt.show();

if __name__ == "__main__":
	parser = argparse.ArgumentParser(
		usage='python {0} datafile.csv -t "title_text"'.format(__file__))
	parser.add_argument("filename", help="Filename of .csv data file")
	parser.add_argument('-t', 
		help = "Graph title text. Uses filename if none.")
	args = parser.parse_args()

	if args.t is None:
		plotBagTest(args.filename, args.filename).run()
	else:
		plotBagTest(args.filename, args.t).run()
	


fs6122.h
#ifndef FS6122_H
#define FS6122_H
#include <Encoder.h>    // for byte

//////////////////////////////////////////////////////
// Siargo Ltd fs6122 sensor
// see httpfps://www.servoflo.com/download-archive/data-sheets/254-mass-flow-vacuum-sensors/1220-fs6122-datasheet
// and https://www.servoflo.com/download-archive/application-notes/212-mass-flow/1256-fs-i2c-communication
// and https://www.digikey.com/product-detail/en/siargo-ltd/FS6122-250F250-0P0-0/2257-FS6122-250F250-0P0-0-ND/11657804?utm_adgroup=Siargo%20Ltd&utm_source=google&utm_medium=cpc&utm_campaign=Shopping_DK%2BSupplier_Other&utm_term=&utm_content=Siargo%20Ltd&gclid=EAIaIQobChMIkJmP45_c6QIVCtvACh1LvwBoEAYYASABEgL9X_D_BwE
#define FS6122_ADDRESS 0x01 //Default
#define FS6122_FILTER_DEPTH 32 //0-128
#define FS6122_CMD_WRITE_FILTER_DEPTH byte(0x0B)
#define FS6122_CMD_CAL_FLOW_RATE byte(0x1C)
#define FS6122_CMD_CAL_PRESSURE byte(0x24)
#define FS6122_CMD_READ_FLOW_RATE byte(0x83)
#define FS6122_CMD_READ_FLOW_RATE_AND_PRESSURE byte(0x84)
#define FS6122_CMD_READ_PRESSURE byte(0xA3)
#define FS6122_CMD_READ_TEMPERATURE byte(0xB2)
#define FS6122_CMD_READ_HUMIDITY byte(0xB3)

typedef struct{
   long umH2O;
   long mSLPM;
} pressure_flow_type;

void fs6122_init();
void fs6122_zeroFlowCal();
void fs6122_readSmlpM_umH2O(pressure_flow_type *pf);
float fs6122_readSmlpM();

#endif // FS6122_H
